File: java-oauth-dpop/src/main/java/com/github/prodnull/oauthdpop/crosstest/CrossTest.java
package com.github.prodnull.oauthdpop.crosstest;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.github.prodnull.oauthdpop.*;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;

public class CrossTest {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final String METHOD = "POST";
    private static final String TARGET = "https://cross-test.example.com/token";

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: CrossTest <generate|validate> [proof_file]");
            System.exit(1);
        }

        try {
            switch (args[0]) {
                case "generate":
                    String outputFile = args.length > 1 ? args[1] : "java_proof.json";
                    generate(outputFile);
                    break;
                case "validate":
                    if (args.length < 2) {
                        System.err.println("Usage: CrossTest validate <proof_file>");
                        System.exit(1);
                    }
                    validate(args[1]);
                    break;
                default:
                    System.err.println("Unknown command: " + args[0]);
                    System.exit(1);
            }
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static void generate(String outputFile) throws Exception {
        DPoPClient client = DPoPClient.generate();
        String proof = client.createProof(METHOD, TARGET, null);

        ObjectNode data = MAPPER.createObjectNode();
        data.put("proof", proof);
        data.put("thumbprint", client.getThumbprint());
        data.put("method", METHOD);
        data.put("target", TARGET);

        MAPPER.writerWithDefaultPrettyPrinter().writeValue(new File(outputFile), data);
        System.out.println("Generated proof: " + outputFile);
    }

    private static void validate(String inputFile) throws Exception {
        String json = Files.readString(Path.of(inputFile));
        var data = MAPPER.readTree(json);

        DPoPConfig config = DPoPConfig.builder()
                .maxProofAgeSecs(300) // 5 minutes for cross-language tests
                .requireNonce(false)
                .expectedMethod(data.get("method").asText())
                .expectedTarget(data.get("target").asText())
                .build();

        try {
            String thumbprint = DPoPValidator.validateProof(data.get("proof").asText(), config);
            DPoPValidator.verifyBinding(thumbprint, data.get("thumbprint").asText());
            System.out.println("PASS: " + inputFile + " validated successfully");
            System.exit(0);
        } catch (DPoPValidationException e) {
            System.err.println("FAIL: " + e.getCode() + ": " + e.getMessage());
            System.exit(1);
        }
    }
}

File: dpop-cross-language-tests/python-test/cross_test.py
#!/usr/bin/env python3
"""Cross-language DPoP test for Python."""

import json
import sys
from pathlib import Path

# Add parent directory to path for local import
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "python-oauth-dpop"))

from oauth_dpop import DPoPClient, DPoPConfig, validate_proof, verify_binding, DPoPValidationError

METHOD = "POST"
TARGET = "https://cross-test.example.com/token"


def generate(output_file: str):
    """Generate a proof and save to file."""
    client = DPoPClient.generate()
    proof = client.create_proof(METHOD, TARGET)

    data = {
        "proof": proof,
        "thumbprint": client.thumbprint,
        "method": METHOD,
        "target": TARGET,
    }

    with open(output_file, "w") as f:
        json.dump(data, f, indent=2)

    print(f"Generated proof: {output_file}")


def validate(input_file: str):
    """Validate a proof from file."""
    with open(input_file) as f:
        data = json.load(f)

    config = DPoPConfig(
        max_proof_age_secs=300,  # 5 minutes for cross-language tests
        require_nonce=False,
        expected_method=data["method"],
        expected_target=data["target"],
    )

    try:
        thumbprint = validate_proof(data["proof"], config)
        verify_binding(thumbprint, data["thumbprint"])
        print(f"PASS: {input_file} validated successfully")
        sys.exit(0)
    except DPoPValidationError as e:
        print(f"FAIL: {e.code}: {e.message}", file=sys.stderr)
        sys.exit(1)


def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <generate|validate> [proof_file]", file=sys.stderr)
        sys.exit(1)

    command = sys.argv[1]

    if command == "generate":
        output_file = sys.argv[2] if len(sys.argv) > 2 else "python_proof.json"
        generate(output_file)
    elif command == "validate":
        if len(sys.argv) < 3:
            print(f"Usage: {sys.argv[0]} validate <proof_file>", file=sys.stderr)
            sys.exit(1)
        validate(sys.argv[2])
    else:
        print(f"Unknown command: {command}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
